api/institutions/authentication.py

```diff
@@ -26,42 +54,50 @@ def authenticate(self, request):
         except (jwt.InvalidTokenError, TypeError):
             raise AuthenticationFailed
 
-        # The JWT `data` payload is expected in the following structure.
-        #
-        # {"provider": {
-        #     "idp": "https://login.circle.edu/idp/shibboleth",
-        #     "id": "CIR",
-        #     "user": {
-        #         "middleNames": "",
-        #         "familyName": "",
-        #         "givenName": "",
-        #         "fullname": "Circle User",
-        #         "suffix": "",
-        #         "username": "user@circle.edu"
-        #     }
-        # }}
         data = json.loads(payload['data'])
         provider = data['provider']
 
         institution = Institution.load(provider['id'])
         if not institution:
             raise AuthenticationFailed('Invalid institution id specified "{}"'.format(provider['id']))
 
-        username = provider['user']['username']
-        fullname = provider['user']['fullname']
+        username = provider['user'].get('username')
+        fullname = provider['user'].get('fullname')
+        given_name = provider['user'].get('givenName')
+        family_name = provider['user'].get('familyName')
+        middle_names = provider['user'].get('middleNames')
+        suffix = provider['user'].get('suffix')
 
-        user, created = get_or_create_user(fullname, username, reset_password=False)
+        # use given name and family name to build full name if not provided
+        if given_name and family_name and not fullname:
+            fullname = given_name + ' ' + family_name
+
+        # institution must provide `fullname`, otherwise we fail the authentication and inform sentry
+        if not fullname:
+            message = 'Institution login failed: fullname required' \
+                      ' for user {} from institution {}'.format(username, provider['id'])
+            sentry.log_message(message)
+            raise AuthenticationFailed(message)
 
+        # `get_or_create_user()` guesses names from fullname
+        # replace the guessed ones if the names are provided from the authentication
+        user, created = get_or_create_user(fullname, username, reset_password=False)
         if created:
-            user.given_name = provider['user'].get('givenName') or user.given_name
-            user.middle_names = provider['user'].get('middleNames') or user.middle_names
-            user.family_name = provider['user'].get('familyName') or user.family_name
-            user.suffix = provider['user'].get('suffix') or user.suffix
+            if given_name:
+                user.given_name = given_name
+            if family_name:
+                user.family_name = family_name
+            if middle_names:
+                user.middle_names = middle_names
+            if suffix:
+                user.suffix = suffix
```

Comment:
`get_or_create_user()` guesses the names based on `fullname`. If these names are provided by the IdP, replace the guessed ones with the provided.